File: prototype/rails/src/test/domain/payments/fake_card_test.rb
require "test_helper"

module Domain
  module Payments
    class FakeCardTest < ActiveSupport::TestCase
      def test_the_approved_number_is_approved
        assert_predicate FakeCard.decide("4242424242424242"), :approved?
      end

      def test_spaces_and_dashes_are_ignored
        assert_predicate FakeCard.decide("4242 4242-4242 4242"), :approved?
      end

      def test_the_generic_decline_number_is_declined
        decision = FakeCard.decide("4000 0000 0000 0002")

        refute_predicate decision, :approved?
        assert_equal DeclineReason::GENERIC_DECLINE, decision.decline_reason
      end

      def test_the_insufficient_funds_number_is_declined
        assert_equal DeclineReason::INSUFFICIENT_FUNDS, FakeCard.decide("4000 0000 0000 9995").decline_reason
      end

      def test_any_other_number_is_not_a_valid_card
        assert_equal DeclineReason::INVALID_CARD_NUMBER, FakeCard.decide("1234 5678 1234 5678").decline_reason
      end

      def test_only_the_last_four_digits_come_back
        assert_equal "9995", FakeCard.decide("4000 0000 0000 9995").last_four
      end

      def test_a_number_shorter_than_four_digits_keeps_what_it_has
        assert_equal "12", FakeCard.decide("12").last_four
      end
    end
  end
end

File: prototype/php/src/app/Domain/Payments/FakeCardTest.php
<?php

declare(strict_types=1);

namespace App\Domain\Payments;

/*
|--------------------------------------------------------------------------
| preg_replace() override
|--------------------------------------------------------------------------
|
| tests/FunctionOverrides.php declares the App\Domain\Payments\preg_replace()
| override this file forces below; see that file for why it lives there
| rather than here.
|
*/

it('treats a preg_replace failure as an invalid, empty-numbered card', function (): void {
    $GLOBALS['fakeCardForcePregReplaceNull'] = true;

    try {
        $decision = FakeCard::decide('4242424242424242');
    } finally {
        $GLOBALS['fakeCardForcePregReplaceNull'] = false;
    }

    expect($decision->isApproved)->toBeFalse()
        ->and($decision->lastFour)->toBe('')
        ->and($decision->declineReason)->toBe(DeclineReason::InvalidCardNumber);
});

it('decides approval or a decline reason from the card number', function (string $number, ?DeclineReason $expected): void {
    $decision = FakeCard::decide($number);

    expect($decision->isApproved)->toBe($expected === null)
        ->and($decision->declineReason)->toBe($expected);
})->with([
    '4242 4242 4242 4242 is approved' => ['4242 4242 4242 4242', null],
    '4000 0000 0000 0002 is declined: generic decline' => ['4000 0000 0000 0002', DeclineReason::GenericDecline],
    '4000 0000 0000 9995 is declined: insufficient funds' => ['4000 0000 0000 9995', DeclineReason::InsufficientFunds],
    'anything else is declined: invalid card number' => ['1234 5678 9012 3456', DeclineReason::InvalidCardNumber],
]);

it('ignores separators between the digits', function (string $number): void {
    expect(FakeCard::decide($number)->isApproved)->toBeTrue();
})->with([
    'dashes' => ['4242-4242-4242-4242'],
    'no separator at all' => ['4242424242424242'],
    'a mix of spaces and dashes, including surrounding whitespace' => ['  4242 4242-4242 4242  '],
]);

it('exposes the last four digits', function (): void {
    expect(FakeCard::decide('4000-0000-0000-9995')->lastFour)->toBe('9995');
});

it('exposes every digit of a number shorter than four', function (): void {
    expect(FakeCard::decide('42')->lastFour)->toBe('42');
});

it('declines an empty number as invalid', function (): void {
    $decision = FakeCard::decide('   ');

    expect($decision->isApproved)->toBeFalse()
        ->and($decision->lastFour)->toBe('')
        ->and($decision->declineReason)->toBe(DeclineReason::InvalidCardNumber);
});
